website/server/libs/taskManager.js

```diff
@@ -281,3 +291,176 @@ export function moveTask (order, taskId, to) {
+        direction,
+      });
+      managerPromises.push(manager.save());
+    });
+
+    managerPromises.push(task.save());
+    await Promise.all(managerPromises);
+
+    throw new NotAuthorized(res.t('taskApprovalHasBeenRequested'));
+  }
+  let wasCompleted = task.completed;
+
+  let [delta] = shared.ops.scoreTask({task, user, direction}, req);
+  // Drop system (don't run on the client, as it would only be discarded since ops are sent to the API, not the results)
+  if (direction === 'up') shared.fns.randomDrop(user, {task, delta}, req, res.analytics);
+
+  // If a todo was completed or uncompleted move it in or out of the user.tasksOrder.todos list
+  // TODO move to common code?
+  let taskOrderPromise;
+  let pullTask = false;
+  let pushTask = false;
+  if (task.type === 'todo') {
+    if (!wasCompleted && task.completed) {
+      // @TODO: mongoose's push and pull should be atomic and help with
+      // our concurrency issues. If not, we need to use this update $pull and $push
+      pullTask = true;
+      // user.tasksOrder.todos.pull(task._id);
+    } else if (wasCompleted && !task.completed && user.tasksOrder.todos.indexOf(task._id) === -1) {
+      pushTask = true;
+      // user.tasksOrder.todos.push(task._id);
+    }
+  }
+
+  setNextDue(task, user);
+
+  if (task.group && task.group.taskId) {
+    await handleSharedCompletion(task);
+  }
+
+  taskScoredWebhook.send(user, {
+    task,
+    direction,
+    delta,
+    user,
+  });
+
+  // Track when new users (first 7 days) score tasks
+  if (moment().diff(user.auth.timestamps.created, 'days') < 7) {
+    res.analytics.track('task score', {
+      uuid: user._id,
+      hitType: 'event',
+      category: 'behavior',
+      taskType: task.type,
+      direction,
+      headers: req.headers,
+    });
+  }
+  return {user, task, delta, direction, taskOrderPromise, pullTask, pushTask, _tmp: _.cloneDeep(user._tmp)};
+}
+
+async function handleChallengeTask (task, delta, direction, resolve, reject) {
+  if (task.challenge && task.challenge.id && task.challenge.taskId && !task.challenge.broken && task.type !== 'reward') {
+    // Wrapping everything in a try/catch block because if an error occurs using `await` it MUST NOT bubble up because the request has already been handled
+    try {
+      const chalTask = await Tasks.Task.findOne({
+        _id: task.challenge.taskId,
+      }).exec();
+
+      if (!chalTask) return;
+
+      await chalTask.scoreChallengeTask(delta, direction);
+    } catch (e) {
+      logger.error(e);
+      reject();
+    }
+  }
+  resolve();
+}
+
+export async function scoreTasks (user, taskScorings, req, res) {
+  let taskIds = taskScorings.map(taskScoring => taskScoring.id).filter(id => id !== undefined);
+  if (taskIds.length === 0) throw new BadRequest(res.t('taskNotFound'));
+  let tasks = {};
+  (await Tasks.Task.findMultipleByIdOrAlias(taskIds, user._id)).forEach(task => {
+    tasks[task._id] = task;
+    if (task.alias && task.alias.length > 0) {
```

Comment:
I still think this will result in tasks with an alias being added twice

Reply:
Yes. They will be added twice to this object. But they will only be processed and saved once. This was needed, since the list it iterates on for scoring the tasks might not have the task ID and only has the alias. Having both the ID and alias available is the easiest way around this. Since the object (should) use references to the same task anyway the overhead should be negligible.